lib/spike.rb

```diff
@@ -0,0 +1,361 @@
+require 'forwardable'
+require 'json'
+require 'pathname'
+require 'active_support/inflector'
+
+module Spike
+  class Resource
+    attr_reader :path, :segments
+    def initialize(path)
+      @path = path.to_s
+      @segments = @path.split("/").reject(&:empty?)
+    end
+
+    def name
+      s = objects.first
+      if subresource?
+        prefix, resource = objects
+        s = [prefix.singularize, resource].join("_")
+      end
+      singular?? s.singularize : s
+    end
+
+    def singular?
+      segments.last =~ /:.*id$/
+    end
+
+    def subresource?
+      objects.count > 1
+    end
+
+    def objects
+      return @objects if defined?(@objects)
+
+      @objects = segments.reject {|segment| segment.start_with?(":")}
+      if @objects.count > 1
+        @objects = @objects.reject {|resource| resource == "repos"}
+      end
+      @objects
+    end
+  end
+
+  class Routes
+    attr_reader :definition
+    def initialize(json)
+      @definition = JSON.parse(json, object_class: OpenStruct)
+      normalize!
+    end
+
+    private
+
+    def normalize!
+      @definition["original_path"] = definition.path
+      @definition["path"]          = normalized_path
```

Comment:
original vs actual path -- for now this is whenever we have something in the path that says `:id` we need it to be prefixed with the actual resource name (singular).